flink-state-backends/flink-statebackend-forst/src/main/java/org/apache/flink/state/forst/fs/filemapping/FileMappingManager.java

```diff
@@ -0,0 +1,252 @@
+ * -> local file) and linked mapping (remote file -> remote file). Note, the key/value of mapping
+ * table must be a file path, directories are maintained by file system itself, directories wouldn't
+ * be the key/value of mapping table.
+ */
+public class FileMappingManager {
+
+    private static final Logger LOG = LoggerFactory.getLogger(FileMappingManager.class);
+
+    public static final String SST_SUFFIX = ".sst";
+
+    private final FileSystem fileSystem;
+
+    private final FileSystem localFileSystem;
+
+    private final HashMap<String, MappingEntry> mappingTable;
+
+    private final String remoteBase;
+
+    private final String localBase;
+
+    public FileMappingManager(
+            FileSystem fileSystem,
+            FileSystem localFileSystem,
+            String remoteBase,
+            String localBase) {
+        this.fileSystem = fileSystem;
+        this.localFileSystem = localFileSystem;
+        this.mappingTable = new HashMap<>();
+        this.remoteBase = remoteBase;
+        this.localBase = localBase;
+    }
+
+    /** Create a mapping entry for a file. */
+    public RealPath createFile(Path file) {
+        String fileName = file.toString();
+        Preconditions.checkState(!mappingTable.containsKey(fileName));
+        if (!fileName.endsWith(SST_SUFFIX) && fileName.startsWith(remoteBase)) {
+            Path localFile = new Path(localBase, file.getName());
+            mappingTable.put(
+                    fileName, new MappingEntry(1, localFileSystem, localFile.toString(), false));
+            return new RealPath(localFile, true);
+        } else {
+            mappingTable.put(fileName, new MappingEntry(1, fileSystem, fileName, false));
+            return new RealPath(file, false);
+        }
+    }
+
+    /** Called by link/copy. Directory link is not supported now. */
+    public int link(String src, String dst) {
+        if (src.equals(dst)) {
+            return -1;
+        }
+        // if dst already exist, not allow
+        if (mappingTable.containsKey(dst)) {
+            return -1;
+        }
+        MappingEntry sourceEntry = mappingTable.get(src);
+        if (sourceEntry != null) {
+            sourceEntry.retain();
+            mappingTable.putIfAbsent(dst, sourceEntry);
+        } else {
+            sourceEntry = new MappingEntry(0, fileSystem, src, false);
+            sourceEntry.retain();
+            mappingTable.put(src, sourceEntry);
+            sourceEntry.retain();
+            mappingTable.put(dst, sourceEntry);
+        }
+        LOG.trace("link: {} -> {}", dst, src);
+        return 0;
+    }
+
+    /** Get the real path of a file, the real path maybe a local file or a remote file/dir. */
+    public RealPath realPath(Path file) {
+        String fileName = file.toString();
+        MappingEntry entry = mappingTable.getOrDefault(fileName, null);
+        if (!fileName.endsWith(SST_SUFFIX) && fileName.startsWith(remoteBase)) {
```

Comment:
Could we determine whether this is a local file via the information in `MappingEntry`? No more branch here.